openedx/core/lib/graph_traversals.py

```diff
@@ -0,0 +1,257 @@
+"""
+This module contains generic generator functions for traversing tree
+(and DAG) structures.  It is agnostic to the underlying data structure
+and implementation of the tree object.  It does this through dependency
+injection of the tree's accessor functions: get_parents and
+get_children.
+
+Additionally, it supports the visitor design pattern by accepting a
+get_result function that is called on each visited node.
+
+The following depth-first traversal methods are implemented:
+
+* Pre-order: Parent yielded before children; child with multiple
+parents is yielded when first encountered.
+
+* Topological: Parent yielded before children; child with multiple
+parents yielded only after all its parents are visited.
+
+* Post-order: Children yielded before its parents.
+
+Note: In-order traversal is not implemented as of yet.  We can do so
+if/when needed.
+
+Optimization once DAGs are not supported:
+Supporting Directed Acyclic Graphs (DAGs) requires us to use
+topological sort, which has the following negative performance
+implications:
+
+* For a simple tree, we can immediately skip over traversing
+descendants, once it is determined that a parent is not to be yielded
+(based on the return value from the 'filter_func' function). However,
+since we support DAGs, we cannot simply skip over descendants since
+they may still be accessible through a different ancestry chain and
+need to be revisited once all their parents are visited.
+
+* For topological sort, we need the get_parents accessor function in
+order to determine whether all of a node's parents have been visited.
+This means the underlying implementation of the graph needs to have
+an efficient way to get a node's parents, perhaps with back pointers
+to each node's parents.  This requires additional storage space, which
+could be eliminated if DAGs are not supported.
+
+
+"""
+from collections import deque
+
+
+def traverse_topologically(
+    start_node, get_parents, get_children, get_result=None, filter_func=None, yield_descendants_of_unyielded=False,
+):
```

Comment:
Not a requirement for this PR, but I am a bit curious why these are all top level functions, instead of a class that gets initialized with `get_parents`, `get_children`, `get_result`, and is queried with different traversals (where  `start_node`, `filter_func`, `yield_descendants_of_unyielded` are passed as params)?